allennlp/models/vilbert.py

```diff
@@ -714,6 +789,95 @@ def __init__(
         self.classifier = torch.nn.Linear(pooled_output_dim * 2, 2)
         self.dropout = torch.nn.Dropout(dropout)
 
+    @classmethod
+    def from_huggingface_model_name(
+        cls,
+        vocab: Vocabulary,
+        model_name: str,
+        image_feature_dim: int,
+        image_num_hidden_layers: int,
+        image_hidden_size: int,
+        combined_hidden_size: int,
+        pooled_output_dim: int,
+        image_intermediate_size: int,
+        image_attention_dropout: float,
+        image_hidden_dropout: float,
+        v_biattention_id: List[int],
+        t_biattention_id: List[int],
+        fixed_t_layer: int,
+        fixed_v_layer: int,
+        pooled_dropout: float = 0.1,
+        fusion_method: str = "sum",
+        fast_mode: bool = False,
+        with_coattention: bool = True,
+        in_batch_pairs: bool = False,
+    ):
+        transformer = AutoModel.from_pretrained(model_name)
+
+        # TODO(mattg): this is hard-coded for BERT; not sure it works for all transformers, or what
+        # to do if it fails.  We should probably pull this out into a central "transformers_util"
+        # module, or something, and just have a method that pulls out an initialized embedding layer
+        # from a huggingface model.
+        text_embeddings = deepcopy(transformer.embeddings)
+        if hasattr(transformer.config, "embedding_size"):
+            config = transformer.config
+
+            from transformers.modeling_albert import AlbertModel
+
+            if isinstance(transformer, AlbertModel):
+                linear_transform = deepcopy(transformer.encoder.embedding_hidden_mapping_in)
+            else:
+                logger.warning(
+                    "Unknown model that uses separate embedding size; weights of the linear "
+                    f"transform will not be initialized.  Model type is: {transformer.__class__}"
+                )
+                linear_transform = torch.nn.Linear(config.embedding_dim, config.hidden_dim)
+
+            # We can't just use torch.nn.Sequential here, even though that's basically all this is,
+            # because Sequential doesn't accept *inputs, only a single argument.
+
+            class EmbeddingsShim(torch.nn.Module):
+                def __init__(self, embeddings: torch.nn.Module, linear_transform: torch.nn.Module):
+                    super().__init__()
+                    self.linear_transform = linear_transform
+                    self.embeddings = embeddings
+
+                def forward(self, *inputs, **kwargs):
+                    return self.linear_transform(self.embeddings(*inputs, **kwargs))
```

Comment:
These embeddings are just a linear transform on top of existing pre-trained embeddings? What's the point of that?

Are the embeddings shared in some other place, and that's why? If these embedding weights are supposed to be tied to some other weights somewhere, I'm pretty sure that the `deepcopy` above breaks that.

Reply:
The pipeline in a transformer is essentially:

```
hidden_states = self.embeddings(inputs)
for layer in self.layers:
    hidden_states = layer(hidden_states)
```

For this logic to work on its own, `embedding_dim` must equal `hidden_dim`, and this is what BERT (and most other transformers) do.  Albert decides to have a much smaller `embedding_dim`, with a linear transform after the embedding to get the hidden states up to the right size:

```
hidden_states = self.linear_transform(self.embeddings(inputs))
for layer in self.layers:
    hidden_states = layer(hidden_states)
```

This reduces the number of parameters in your model: embedding matrix goes from `(vocab_size, 768)` to `(vocab_size, 128)`, while you add a `(128, 768)` linear transform matrix.

There is nothing that's supposed to be tied anywhere; I used `deepcopy` here to get a new copy of these existing weight matrices (both the embeddings and the linear transform), to initialize the weights in our model correctly.  Duplicating the modules themselves is much easier than aligning all of the underlying `Parameters` and calling `.copy_()`.

Reply:
Hmm, it's possible there is an activation in between the embeddings and the transform in albert; I should double check that.

Reply:
Makes sense :+1:

Reply:
I just checked, and there is no activation in between the embedding and the hidden layer in albert, so this logic is correct.